wallet: allow skipping script paths

Sjors opened the pull request:
While testing MuSig2 integration in #29675 I ran into a Ledger bug https://github.com/LedgerHQ/app-bitcoin-new/issues/329 that was triggered because we signed for a script path in addition to the key path. The bug itself will be fixed, but it seems useful in general to be able to skip script path signing.

For example, you may have a 2-of-2 musig() descriptor with some sort of fallback after coins are N blocks old. For both on chain privacy and fee reasons, you don't want to use this path unless absolutely necessary. But our wallet can't know if our co-signer still has their key, so we sign the script path at all times. This could lead to accidental broadcast.

This PR introduces a `keypath_only` option (default `false` for now) to the following RPC methods:
- `send` 
- `sendall`
- `walletprocesspsbt`
- `descriptorprocesspsbt`

The resulting PSBT won't _add_ `taproot_scripts` or `taproot_script_path_sigs`. It does still have `taproot_bip32_derivs` and `musig2_participant_pubkeys`.

It will not remove `taproot_scripts` from an existing PSBT, but it won't add a corresponding entry to `taproot_script_path_sigs`.

The `wallet_taproot.py` tests are expanded to cover this behavior.

It can also be tested with https://github.com/Sjors/bitcoin/pull/91.

Sjors committed:
test: cover keypath_only in wallet_taproot.py

Expand taproot tests to cover avoid_script_path in  walletprocesspsbt.

When avoiding script paths, there's no need for the workaround that increases fee_rate to compensate for the wallet's inability to estimate fees for script path spends. We use this to indirectly test that key path was used.

We also check that taproot_script_path_sigs is not set.

Finally, for transactions that can't be signed using their key path, we try again by allowing the script path.

Additional test extended private keys were extracted from other tests.

Co-authored-by: rkrux <rkrux.connect@gmail.com>

Sjors committed:
rpc: add keypath_only to walletprocesspsbt

Sjors committed:
rpc: add keypath_only to send and sendall

Sjors commented:
The `walletprocesspsbt` RPC now also has this option. Added test coverage by expanding `wallet_taproot.py`.

Sjors commented:
I split off a non-behavior-changing commit that adds `avoid_script_path` to `FillPSBT`. I think it makes sense to pass this option here, although in general we may want to refactor `FillPSBT` to an options struct instead of this ever expanding argument list. That could be an independent or prerequisite PR.

I then did the same for `SignPSBTInput`, where I also think struct would be better.

Rinse and repeat for `::SignTransaction`.

Will continue later to further split 09334ae0bd7939ed7476eb2b36f3bf8085885fe7. It might make sense to add a property to `BaseSignatureCreator` or it subclasses?

Sjors committed:
wallet: add option to avoid script path spends

Sjors commented:
I expanded #32876 with `SignOptions`. This reduces e2a151c3dc35997ac0f515feb292fc4543093593 to a very small ... but very ugly commit: casting `BaseSignatureCreator` to `MutableTransactionSignatureCreator` to access its `avoid_script_path` option.

Sjors commented:
This may also be useful for Silent Payments, cc @josibake.

Sjors commented:
Sidenote: currently it's unlikely you'll accidentally send an `older()` or `after()` script path. This is because you need to manually set the input `sequence` field in the `send` (etc.) RPC (and manually adjust the fee rate). Found out the hard way while testing 🤦 (there's currently no useful feedback when this happens).

rkrux commented:
> Sidenote: currently it's unlikely you'll accidentally spend via an `older()` or `after()` script path. This is because you need to manually set the input `sequence` field in the `send` (etc.) RPC (and manually adjust the fee rate). Found out the hard way while testing 🤦 (there's currently no useful feedback when this happens).

**Scenario that I have not tested but expect to behave in this way:** 3 signers (A, B, C) form a MuSig (ABC) in the `keypath` and also have 3 non-Musig multi-sig script paths (AB, BC, AC) without any timelocks. They intend to MuSig sign in order from A to C to spend via keypath and start the signing flow first by producing nonces in the mentioned order, and then by adding partial signatures. By the time B adds its nonce, there'd would be a fully signed spendable script path? If yes, seems unusual to have this when the intent of the signers was for keypath spending, which makes a good case for this PR.

I did notice unconditional key and script path signing within `SignTaproot` function when I was reviewing #29675. We tried to simulate an intentional script path spending like above in #29675 by making `A` a no-signing wallet in the last test case in functional tests: https://github.com/bitcoin/bitcoin/pull/29675/commits/ac599c4a9cb3b2d424932d3fd91f9eed17426827#diff-c94f9555a2569240d362a00a82764f4714f6c65283e3f96725cf2a1043a296b5R238. Though the script path in that case is using MuSig but the overall unconditional script signing principle is still applicable.

rkrux commented:
If #32876 is not garnering reviews, can consider undrafting this PR itself because it contains 7 commits only. 
Allowing skipping script path spending optionally seems reasonable to me and a ready-to-review PR is more inviting than a draft one.

Sjors commented:
@rkrux I'm not too worried about reviews just yet, let's focus on getting  https://github.com/bitcoin/bitcoin/pull/29675 in.

Sjors commented:
Rebased after #32876 landed. I need to check a few more things before marking this as ready for review.

Sjors commented:
Rebased after #29136 landed.

Added test coverage for the `send` RPC (ad364249baebe9f75f24e32358d1f1f514c98f1d -> 8bd15fd779f8ac43c15f32d8b52a6439ec738d9f)

Redesigned ed89145afb12138b8e6664bc9acdda4772028115 _wallet: add option to avoid script path spends_ to pass `SignOptions` through `ProduceSignature` into `SignTaproot` (9bfdfaea2dbd62b8f86be90df18fb9cd61bac60f + f77a74743ececdf98dff9188c19f13c725b21801).

`SignTaproot` now also clears `spenddata.scripts` when `avoid_script_path` is set.

rkrux commented on src/rpc/client.cpp:
```diff
@@ -214,6 +214,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "walletprocesspsbt", 2, "sighashtype", ParamFormat::STRING },
     { "walletprocesspsbt", 3, "bip32derivs" },
     { "walletprocesspsbt", 4, "finalize" },
+    { "walletprocesspsbt", 5, "keypath_only"},
     { "descriptorprocesspsbt", 0, "psbt", ParamFormat::STRING },
```
In 52a5ace3718dae474ce09c1e06aa26293a4e82dc "rpc: add keypath_only to walletprocesspsbt"

descriptorprocesspsbt also does signing, it won't be far fetched to allow this option in that RPC too.

Sjors replied:
Done in 8d819f7d8e6e93b8132ae0e621ad39740a9493df.

rkrux commented on src/rpc/client.cpp:
```diff
@@ -267,6 +267,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "send", 4, "replaceable"},
     { "send", 4, "solving_data"},
     { "send", 4, "max_tx_weight"},
+    { "send", 4, "keypath_only"},
     { "send", 5, "version"},
     { "sendall", 0, "recipients" },
```
In 8bd15fd779f8ac43c15f32d8b52a6439ec738d9f "rpc: add keypath_only to send"

FinishTransaction is updated in this PR to have this option. It's used by send and sendall, so it seems reasonable to have this option in sendall too.

rkrux replied:
In 8bd15fd "rpc: add keypath_only to send"

Along similar lines, sendmany and sendtoaddress also have similar expectations as send, which includes signing. Although it seems congruent to add this option in those RPCs too but it might involve a little bit more refactoring because of a slightly different flow taken by them. Though ultimately they end up calling SignTransaction that already has been updated in this PR.

Sjors replied:
I'll add `sendall`, but I'm inclined leave the older `sendmany` and `sendtoaddress` alone. They also can't be used with external signers.

rkrux replied:
Oh, I didn't know they were older.

Seems fine to me to not update them in this PR, they don't use the common internal signing/processing functions as well.

rkrux reviewed (commented):
Concept ACK 8bd15fd779f8ac43c15f32d8b52a6439ec738d9f

Seems like a useful feature to have, the example in the PR description is helpful for the motivation. Asked a couple questions around updating other similar RPCs.

rkrux commented on test/functional/wallet_taproot.py:
```diff
@@ -384,10 +459,12 @@ def do_test_psbt(self, comment, pattern, privmap, treefn, keys_pay, keys_change)
-        self.do_test_psbt(comment, pattern, privmap, treefn, keys[2*nkeys:3*nkeys], keys[3*nkeys:4*nkeys])
+        self.do_test_psbt(comment, pattern, privmap, treefn, keys[2*nkeys:3*nkeys], keys[3*nkeys:4*nkeys], avoid_script_path=False)
+        if 'tr' in pattern:
+            self.do_test_psbt(comment, pattern, privmap, treefn, keys[4*nkeys:5*nkeys], keys[5*nkeys:6*nkeys], avoid_script_path=True)
```
In fec2019bdb3f97482fba54a1a4540f9e4ec44efd "test: cover keypath_only in wallet_taproot.py"

https://github.com/bitcoin/bitcoin/blob/5f33da9aa30ff80876f2f450b4370c446555bd71/test/functional/wallet_musig.py#L352

There's a very interesting test case in wallet_musig.py here where we contrive it to not be able to sign the 3-key-musig-keypath because of nosign_wallets=[0]. A new test case can be added there as well by not passing the nosign option and by instead passing this new option to not sign the script path. This test automatically generates all the required signers and is quite extensible. It would be nice to test the MuSig flow in this PR as well as this use case has been given as an example in the PR description too.

A diff like the below one works. I could not find a quick way to determine if the musig placeholders are within the script path so as to not count them in expected pubnonces and partialsigs.

```diff
diff --git a/test/functional/wallet_musig.py b/test/functional/wallet_musig.py
index 0e3e4377c0..43c2de0436 100755
--- a/test/functional/wallet_musig.py
+++ b/test/functional/wallet_musig.py
@@ -168,7 +168,7 @@ class WalletMuSigTest(BitcoinTestFramework):
         assert "musig2_pubnonces" in dec["inputs"][0]
         assert "musig2_partial_sigs" not in dec["inputs"][0]
 
-    def test_success_case(self, comment, pattern, sighash_type=None, scriptpath=False, nosign_wallets=None, only_one_musig_wallet=False):
+    def test_success_case(self, comment, pattern, sighash_type=None, scriptpath=None, nosign_wallets=None, only_one_musig_wallet=False, expected_pubnonces_partialsigs=None):
         self.log.info(f"Testing {comment}")
         has_internal = MULTIPATH_TWO_RE.search(pattern) is not None
 
@@ -176,22 +176,26 @@ class WalletMuSigTest(BitcoinTestFramework):
         wallets, keys = self.create_wallets_and_keys_from_pattern(pat)
         self.construct_and_import_musig_descriptor_in_wallets(pat, wallets, keys, only_one_musig_wallet)
 
-        expected_pubnonces = 0
-        expected_partial_sigs = 0
-        for musig in MUSIG_RE.findall(pat):
-            musig_partial_sigs = 0
-            for placeholder in PLACEHOLDER_RE.findall(musig):
-                wallet_index = int(placeholder[1:])
-                if nosign_wallets is None or wallet_index not in nosign_wallets:
-                    expected_pubnonces += 1
-                else:
-                    musig_partial_sigs = None
+        if expected_pubnonces_partialsigs != None:
+            expected_pubnonces = expected_pubnonces_partialsigs['pubnonces']
+            expected_partial_sigs = expected_pubnonces_partialsigs['partialsigs']
+        else:
+            expected_pubnonces = 0
+            expected_partial_sigs = 0
+            for musig in MUSIG_RE.findall(pat):
+                musig_partial_sigs = 0
+                for placeholder in PLACEHOLDER_RE.findall(musig):
+                    wallet_index = int(placeholder[1:])
+                    if nosign_wallets is None or wallet_index not in nosign_wallets:
+                        expected_pubnonces += 1
+                    else:
+                        musig_partial_sigs = None
+                    if musig_partial_sigs is not None:
+                        musig_partial_sigs += 1
+                    if wallet_index < len(wallets):
+                        continue
                 if musig_partial_sigs is not None:
-                    musig_partial_sigs += 1
-                if wallet_index < len(wallets):
-                    continue
-            if musig_partial_sigs is not None:
-                expected_partial_sigs += musig_partial_sigs
+                    expected_partial_sigs += musig_partial_sigs
 
         # Check that the wallets agree on the same musig address
         addr = None
@@ -252,7 +256,10 @@ class WalletMuSigTest(BitcoinTestFramework):
         for i, wallet in enumerate(wallets):
             if nosign_wallets and i in nosign_wallets:
                 continue
-            proc = wallet.walletprocesspsbt(psbt=psbt, sighashtype=sighash_type)
+            if scriptpath == None:
+                proc = wallet.walletprocesspsbt(psbt=psbt, sighashtype=sighash_type)
+            else:
+                proc = wallet.walletprocesspsbt(psbt=psbt, sighashtype=sighash_type, keypath_only=not scriptpath)
             assert_equal(proc["complete"], False)
             nonce_psbts.append(proc["psbt"])
 
@@ -280,7 +287,10 @@ class WalletMuSigTest(BitcoinTestFramework):
         for i, wallet in enumerate(wallets):
             if nosign_wallets and i in nosign_wallets:
                 continue
-            proc = wallet.walletprocesspsbt(psbt=comb_nonce_psbt, sighashtype=sighash_type)
+            if scriptpath == None:
+                proc = wallet.walletprocesspsbt(psbt=comb_nonce_psbt, sighashtype=sighash_type)
+            else:
+                proc = wallet.walletprocesspsbt(psbt=comb_nonce_psbt, sighashtype=sighash_type, keypath_only=not scriptpath)
             assert_equal(proc["complete"], False)
             psig_psbts.append(proc["psbt"])
 
@@ -333,6 +343,7 @@ class WalletMuSigTest(BitcoinTestFramework):
         self.test_success_case("tr(H,{pk(musig/*), pk(same keys different musig/*)})", "tr($H,{pk(musig($0,$1,$2)/<0;1>/*),pk(musig($1,$2)/0/*)})", scriptpath=True)
         self.test_success_case("tr(musig/*,{pk(partial keys diff musig-1/*),pk(partial keys diff musig-2/*)})}", "tr(musig($0,$1,$2)/<3;4>/*,{pk(musig($0,$1)/<5;6>/*),pk(musig($1,$2)/7/*)})")
         self.test_success_case("tr(musig/*,{pk(partial keys diff musig-1/*),pk(partial keys diff musig-2/*)})} script-path", "tr(musig($0,$1,$2)/<3;4>/*,{pk(musig($0,$1)/<5;6>/*),pk(musig($1,$2)/7/*)})", scriptpath=True, nosign_wallets=[0])
+        self.test_success_case("tr(musig/*,{pk(partial keys diff musig-1/*),pk(partial keys diff musig-2/*)})} script-path", "tr(musig($0,$1,$2)/<3;4>/*,{pk(musig($0,$1)/<5;6>/*),pk(musig($1,$2)/7/*)})", scriptpath=False, expected_pubnonces_partialsigs={'pubnonces':3, 'partialsigs':3})
         self.test_success_case("tr(H,and(pk(musig/*),after(1)))", "tr($H,and_v(v:pk(musig($0,$1,$2)/<0;1>/*),after(1)))", scriptpath=True)
         self.test_success_case("tr(H,and(pk_k(musig/*),after(1)))", "tr($H,and_v(vc:pk_k(musig($0,$1,$2)/<0;1>/*),after(1)))", scriptpath=True)
         self.test_success_case("tr(H,and(pkh(musig/*),after(1)))", "tr($H,and_v(v:pkh(musig($0,$1,$2)/<0;1>/*),after(1)))", scriptpath=True)

```

Sjors replied:
Taken with some tweaks for readability in 349462b7a97adc221fb2a1d6948f9197e4a1fef1.

Sjors replied:
Touching that file requires a rebase, so will do that now.

rkrux replied:
> Taken with some tweaks for readabilityin https://github.com/bitcoin/bitcoin/commit/349462b7a97adc221fb2a1d6948f9197e4a1fef1.

Nice, the changes in the commit are cleaner than what I suggested.

rkrux reviewed (commented):
The test reminded me of wallet_musig.py functional test that I believe can also be updated here because it works over walletprocesspsbt RPC.

Sjors committed:
rpc: add keypath_only to descriptorprocesspsbt

Sjors committed:
test: cover keypath_only in wallet_musig.py

Extract the descriptor with both MuSig key and script paths into
key_and_script_path_musigs. Reuse it for the existing cases and
give those cases descriptions.

Expand the same scenario with keypath_only=true. The new case asks
walletprocesspsbt to avoid script-path signing.

Co-authored-by: rkrux <rkrux.connect@gmail.com>

Sjors commented:
Rebased after #33636, because the test I added in my previous push touches the same file. Also further simplified that test.

w0xlt commented:
Concept ACK

rkrux reviewed (commented):
I believe the changes in the multiple RPCs require a release note.

Sjors committed:
doc: add release note for keypath_only

Sjors commented:
@rkrux pushed a release note commit.

rkrux commented on test/functional/rpc_psbt.py:
```diff
@@ -1410,6 +1410,36 @@ def test_psbt_input_keys(psbt_input, keys):
+            keypath_only=True,
+        )
+        decoded = self.nodes[2].decodepsbt(keypath_only_psbt["psbt"])
+        assert "taproot_scripts" in decoded["inputs"][0]
```
In 2d8d02d4da89b5e928dd7007365f31ab2f8f9761 "rpc: add keypath_only to descriptorprocesspsbt"

From the PR description:
>  The resulting PSBT won't include taproot_scripts or taproot_script_path_sigs.

The PR description is outdated?

Sjors replied:
When the fields are present, we should not sign them, that was already correct. When the fields are absent, we should not add them. This prevents other signers from accidentally signing them (or get confused by them, as I found with earlier Ledger firmware).

(update: it worked correctly, but PR description was wrong)

I also added some code comments.

rkrux commented on src/common/types.h:
```diff
@@ -49,6 +49,11 @@ struct PSBTFillOptions {
+    /**
+     * Only sign the key path (for taproot inputs).
+     */
+    bool avoid_script_path{false};
```
In fa971a09d5227d29aea57836d46538d6e11ea911 "wallet: add option to avoid script path spends"

I am not seeing any benefit of having two kinds of naming for the same feature - keypath_only in the RPCs and avoid_script_path within the code.

Would really prefer to have only one kind of naming, I'd let keypath_only stay.

Sjors replied:
Good point, I'll switch over to `keypath_only`.

rkrux commented on src/common/types.h:
```diff
@@ -49,6 +49,11 @@ struct PSBTFillOptions {
+    /**
+     * Only sign the key path (for taproot inputs).
+     */
+    bool avoid_script_path{false};
```
In fa971a0 "wallet: add option to avoid script path spends"

PSBTFillOptions is a more generic struct and not having taproot in the avoid_script_path name can be confusing for the uninitiated (though its present in the documentation comment). Maybe prefix it with `taproot_`?

rkrux commented on src/script/sign.h:
In fa971a0 "wallet: add option to avoid script path spends"

There are 2 private and public sections each in the class. Is this intentional? Or can these be clubbed?

Sjors replied:
It is, but it's ugly. I'm adding a `TaprootKeypathOnly()` helper so we can keep `m_options` private and drop the annotations.

Sjors replied:
On second thought, exposing `Options()` seems generic.

rkrux commented on src/script/sign.h:
```diff
@@ -33,6 +33,7 @@ struct SignatureData;
 
 struct SignOptions {
     int sighash_type{SIGHASH_DEFAULT};
+    bool avoid_script_path{false};
 };
```
In fa971a0 "wallet: add option to avoid script path spends"

Same taproot_ prefix suggestion for this generic SignOptions struct as well.

rkrux commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1221,6 +1224,7 @@ RPCMethod send()
                           }},
                         },
                     },
+                    {"keypath_only", RPCArg::Type::BOOL, RPCArg::Default{false}, "Only sign the key path (for taproot inputs)."},
```
In multiple commits:

The false default for all this option in all the affected RPCs can be a wallet constant named DEFAULT_SIGN_TAPROOT_KEYPATH_ONLY.  It needs to be in the wallet namespace, either in this file wallet/rpc/spend.cpp where its used almost exclusively, or in the wallet/wallet.h file where other defaults also are (not my preference though).

Sjors replied:
I'll move it to `wallet/wallet.h` since one day the GUI might have a similar toggle.

`descriptorprocesspsbt` keeps a hardcoded `{false}`, since it doesn't include the wallet.

Ultimately it doesn't matter that much, because changing the default would be a breaking change for older RPC clients, so it seems unlikely to happen.

rkrux commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1644,11 +1645,13 @@ RPCMethod walletprocesspsbt()
     if (sign) EnsureWalletIsUnlocked(*pwallet);
 
-    const auto err{wallet.FillPSBT(psbtx, {.sign = sign, .sighash_type = nHashType, .finalize = finalize, .bip32_derivs = bip32derivs}, complete)};
+    const auto err{wallet.FillPSBT(psbtx, {.sign = sign, .sighash_type = nHashType, .finalize = finalize, .bip32_derivs = bip32derivs, .avoid_script_path = keypath_only}, complete)};
```
In e6b5cef884b9b728954b4e3d3a483b5d41132066 "rpc: add keypath_only to walletprocesspsbt"

Related to an earlier suggestion, seeing this double naming together seems unnecessary to me.

rkrux commented on test/functional/wallet_taproot.py:
```diff
@@ -154,6 +154,56 @@
+        "xprv": "tprv8ZgxMBicQKsPd7Uf69XL1XwhmjHopUGep8GuEiJDZmbQz6o58LninorQAfcKZWARbtRtfnLcJ5MQ2AtHcQJCCRUcMRvmDUjyEmNUWwx8UbK",
+        "xpub": "tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B",
+        "pubs": [
+            "146846eeb5a7533abb594ba734bc243fc7b6349499b8311c8fc13b0112ba8a77",
```
From the comment few lines above:

> xprvs/xpubs, and m/* derived x-only pubkeys (created using independent implementation) 
 
1. Where were these xprvs and xpubs gotten from?
2. Do the 4 pubs here correspond to m/0, m/1, m/2, m/3 respectively?

I don't prefer to have hardcoded extended keys that are created via some outside implementation in the testing framework because it clutters the testing code, is not self documenting, and makes the tests difficult to update - that's why I raised #35543 recently.

If it gets merged before this PR (this one doesn't need to be dependent on it), it'd be really nice to use the ExtendedPrivate(|Public)Key class(es) from there so that we can have programmatic test cases generation that would make these tests self-documenting.

Sjors replied:
1. Lifted from various other tests. Happy to rebase after #35543 and get them from a more canonical source.
2. I believe so, yes.

rkrux replied:
PR #35543 is merged now, only the following change would be required now I believe:

```diff
diff --git a/test/functional/wallet_taproot.py b/test/functional/wallet_taproot.py
index b312d4e153..89d45f0ca5 100755
--- a/test/functional/wallet_taproot.py
+++ b/test/functional/wallet_taproot.py
@@ -256,7 +256,7 @@ class WalletTaprootTest(BitcoinTestFramework):
         self.do_test_psbt(comment, pattern, privmap, treefn, keys[2*nkeys:3*nkeys], keys[3*nkeys:4*nkeys])
 
     def generate_test_keys(self):
-        xprvs = [ExtendedPrivateKey.generate() for _ in range(0, 13)]
+        xprvs = [ExtendedPrivateKey.generate() for _ in range(0, 18)]
         return [{
             "xprv": xprv.to_string(),
             "xpub": xprv.pubkey().to_string(),
```

Sjors replied:
Done!

rkrux reviewed (commented):
Code review 03541dff2424352b8e366c8deb25ed36190388e0

I'm reviewing the wallet_taproot changes at the moment. I've shared the points that I have till now.

Also the PR description can be updated to mention the remaining changed RPCs.

rkrux commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1644,11 +1645,13 @@ RPCMethod walletprocesspsbt()
     bool sign = request.params[1].isNull() ? true : request.params[1].get_bool();
     bool bip32derivs = request.params[3].isNull() ? true : request.params[3].get_bool();
     bool finalize = request.params[4].isNull() ? true : request.params[4].get_bool();
+    bool keypath_only{request.params[5].isNull() ? false : request.params[5].get_bool()};
```
In e6b5cef "rpc: add keypath_only to walletprocesspsbt"

This point is specific to walletprocesspsbt RPC only because all the 3 other affected RPCs don't provide an option to not sign.

Having both sign and keypath_only booleans might be confusing for the end user as the latter is a form of former. I was reviewing #35370 and noticed multiple overlapping boolean RPC options over there and believe the same point is applicable here as well (to some extent).

One way could be to make this sign option an enum (ALL, NONE, KEYPATH_ONLY), though that would be a breaking change for the end users. An alternative could be to add a condition in the code that sign needs to be true if keypath_only is passed true because keypath_only:true with sign:false is conflicting.

Ideally I prefer the enum approach because it allows us to have fewer RPC arguments but given that it'd be a breaking change and only one RPC needs to be updated at the moment, I'm inclining towards the latter.

WDYT?

Sjors replied:
I think it's better to keep similar RPC methods consistent, even if it's a bit awkward for one of them.

Sjors commented:
Addressed feedback from @rkrux. Most importantly:
- added `Options()` to `BaseSignatureCreator` to avoid having to make `m_options` public. See https://github.com/bitcoin/bitcoin/pull/32857#discussion_r3442943643

Clarified in PR description as well code comments that we don't _add_  new `taproot_scripts`. We also  don't _remove_ exiting ones, but don't _sign_ them. See https://github.com/bitcoin/bitcoin/pull/32857#discussion_r3442425436


Rebased just in case after #35504.

rkrux commented on test/functional/wallet_taproot.py:
It's not related to this PR but I noticed this dead code while reviewing. It can be removed here because this file is being touched, I'm not sure if raising a separate PR only for this would be worth a whole review & merge process.

```diff
--- a/test/functional/wallet_taproot.py
+++ b/test/functional/wallet_taproot.py
@@ -64,9 +64,6 @@ class WalletTaprootTest(BitcoinTestFramework):
     def setup_network(self):
         self.setup_nodes()
 
-    def init_wallet(self, *, node):
-        pass
-
     @staticmethod
     def make_desc(pattern, privmap, keys, pub_only = False):
         pat = pattern.replace("$H", H_POINT)
```

Sjors replied:
@rkrux I'll add a commit here, but it's probably worth its own PR, because it might sit here for a while otherwise.

rkrux replied:
I see, I can raise a separate PR for it then.

rkrux replied:
#35619

Sjors committed:
test: remove unused wallet_taproot init_wallet

Sjors commented:
Rebased after #35543 and took advantage of it in the test, see https://github.com/bitcoin/bitcoin/pull/32857#discussion_r3485600298.

Also added the minor test cleanup suggested in https://github.com/bitcoin/bitcoin/pull/32857#discussion_r3485731084.

Sjors commented:
Rebased after #35852.

jeanpablojp commented on src/script/sign.cpp:
```diff
@@ -612,6 +616,10 @@ static bool SignTaproot(const SigningProvider& provider, const BaseSignatureCrea
 
+    // Key path signing failed. In keypath-only mode, stop here instead of
+    // attempting a script path signature.
+    if (taproot_keypath_only) return false;
```
This return also blocks finalizing via script path with signatures already in the PSBT, not just signing new ones. With `keypath_only=true` and `finalize=true`, such a PSBT comes back `complete=false`. With the flag off it finalizes, and `finalizepsbt` afterwards also works. Probably a rare case, feel free to ignore.

Sjors replied:
Clarified the documentation and added test coverage.

I see `keypath_only=true` as a privacy / safety feature, so this behavior seems correct, even if surprising.

I looked into having the RPC emit a warning, and also enhance `analyzepsbt`, but that's quite involved: https://github.com/Sjors/bitcoin/commits/2026/08/taproot-spend-path-warning/

jeanpablojp commented on src/wallet/rpc/spend.cpp:
```diff
@@ -499,6 +501,7 @@ CreatedTransactionResult FundTransaction(CWallet& wallet, const CMutableTransact
                     {"includeWatching", UniValueType(UniValue::VBOOL)},
                     {"include_watching", UniValueType(UniValue::VBOOL)},
                     {"inputs", UniValueType(UniValue::VARR)},
+                    {"keypath_only", UniValueType(UniValue::VBOOL)},
```
Small note. This list is shared with fundrawtransaction and walletcreatefundedpsbt, which never sign, so both now accept keypath_only with no effect (before, they rejected it). Intentional?

Sjors replied:
I ended up adding support to `walletcreatefundedpsbt`, rather than suppressing the argument.

`fundrawtransaction` now throws if you set it (a legacy method that I prefer to leave alone)

jeanpablojp reviewed (commented):
tACK e730fc4c282876a085c56c28c69103d97772cf85.

Ran the 3 tests this PR touches, all pass. Skipping the fallback script path in the normal flow makes sense. Key path signing still works with the tweak, and script path fields already in the PSBT are left untouched.

Sjors commented:
Rebased after #32958.

Sjors commented:
Rebased after #34075.

Sjors commented:
Rebased after #34697

dangervslash commented:
Resolved
